pkg/netstore/netstore.go

```diff
@@ -47,17 +59,34 @@ func (s *store) Get(ctx context.Context, mode storage.ModeGet, addr swarm.Addres
 				if s.recoveryCallback == nil {
 					return nil, err
 				}
-				targets, err := sctx.GetTargets(ctx)
+				targets, err1 := sctx.GetTargets(ctx)
+				if err1 != nil {
+					return nil, err1
+				}
+				chunkC := make(chan swarm.Chunk, 1)
+				socAddress, err1 := s.recoveryCallback(ctx, addr, targets, chunkC)
+				if err1 != nil {
+					return nil, err1
+				}
+
+				// add the expected soc address and the channel in which the chunk is awaited
+				s.chunkChanMu.Lock()
+				s.chunkChanMap[socAddress.String()] = chunkC
+				s.chunkChanMu.Unlock()
+				defer delete(s.chunkChanMap, socAddress.String())
+
+				// wait for repair response or context to be cancelled
+				var recoveryResponse swarm.Chunk
+				select {
+				case recoveryResponse = <-chunkC:
+				case <-ctx.Done():
+					return nil, err // return the original retrieval error
+				}
+				s, err := soc.FromChunk(recoveryResponse)
 				if err != nil {
-					return nil, err
+					return nil, errDecodingRecoveryResponse
```

Comment:
The actual error is not propagated, it is shadowed with errDecodingRecoveryResponse, which is actually not checked anywhere.